pkg/oauthclientscontroller/oauthclients.go

```diff
@@ -0,0 +1,98 @@
+package oauthclientscontroller
+
+import (
+	"k8s.io/apimachinery/pkg/api/equality"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/klog"
+
+	oauthv1 "github.com/openshift/api/oauth/v1"
+	routev1 "github.com/openshift/api/route/v1"
+	oauthclient "github.com/openshift/client-go/oauth/clientset/versioned/typed/oauth/v1"
+	oauthdiscovery "github.com/openshift/library-go/pkg/oauth/oauthdiscovery"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/resource/resourceapply"
+	"github.com/openshift/library-go/pkg/operator/resource/resourcemerge"
+
+	authoputil "github.com/openshift/cluster-authentication-operator/pkg/util"
+)
+
+const browserClientSecretByteLen = 32
+
+func defaultBrowserOAuthClient(route *routev1.Route) *oauthv1.OAuthClient {
+	return &oauthv1.OAuthClient{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: authoputil.OAuthBrowserClientName,
+		},
+		GrantMethod:  oauthv1.GrantHandlerAuto,
+		Secret:       authoputil.RandomString(browserClientSecretByteLen),
+		RedirectURIs: []string{oauthdiscovery.OpenShiftOAuthTokenDisplayURL(route.Spec.Host)},
+	}
+}
+
+func defaultChallengingOAuthClient(route *routev1.Route) *oauthv1.OAuthClient {
+	return &oauthv1.OAuthClient{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: authoputil.OAuthChallengingClientName,
+		},
+		GrantMethod:           oauthv1.GrantHandlerAuto,
+		RespondWithChallenges: true,
+		RedirectURIs:          []string{oauthdiscovery.OpenShiftOAuthTokenImplicitURL(route.Spec.Host)},
+	}
+}
+
+func (c *oauthClientsController) ensureBrowserOAuthClient(current *oauthv1.OAuthClient, route *routev1.Route) error {
+	defaultBrowserClient := defaultBrowserOAuthClient(route)
+
+	// secret is set properly, don't try to change it
+	if current != nil && len(current.Secret) == browserClientSecretByteLen {
+		defaultBrowserClient.Secret = current.Secret
+	}
+
+	_, _, err := ApplyOAuthClient(c.oauthClientGetter, nil, defaultBrowserClient)
+	return err
+}
+
+func (c *oauthClientsController) ensureChallengingOAuthClient(route *routev1.Route) error {
+	_, _, err := ApplyOAuthClient(c.oauthClientGetter, nil, defaultChallengingOAuthClient(route))
+	return err
+}
+
+// ApplyOAuthClient merges objectmeta. It returns the final Object, whether any change as made, and an error
+func ApplyOAuthClient(
```

Comment:
This is written so that it matches the library-go functions (that's why `recorder` is unused - the functions that would leverage it are not exported in the lib).

I guess I should move this to the appropriate lib if we decide to use it.